File: 9Feb2024.py
from ast import List


class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        maxi = 1
        num = -1
        v = []
        nums.sort()
        dp = [1] * n

        for i in range(1, n):
            for j in range(i):
                if nums[i] % nums[j] == 0 and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
                    if maxi < dp[i]:
                        maxi = dp[i]

        for i in range(n - 1, -1, -1):
            if maxi == dp[i] and (num == -1 or (num % nums[i]) == 0):
                v.append(nums[i])
                maxi -= 1
                num = nums[i]

        return v

File: 9Feb2024.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int n = nums.length;
        int maxi = 1;
        int num = -1;
        List<Integer> v = new ArrayList<>();
        Arrays.sort(nums);
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if ((nums[i] % nums[j]) == 0 && dp[i] < dp[j] + 1) {
                    dp[i] = dp[j] + 1;
                    if (maxi < dp[i]) {
                        maxi = dp[i];
                    }
                }
            }
        }
        for (int i = n - 1; i >= 0; i--) {
            if (maxi == dp[i] && (num == -1 || (num % nums[i]) == 0)) {
                v.add(nums[i]);
                maxi--;
                num = nums[i];
            }
        }
        return v;
    }
}